Policy validation causes a tsunami of errors

If an invalid policy is registered in Nomad, the Autoscaler will output a large amount of error messages in quick succession:

```
2020-06-10T18:13:18.435-0400 [INFO]  agent.plugin_manager: successfully launched and dispensed plugin: plugin_name=nomad-apm
2020-06-10T18:13:18.435-0400 [INFO]  agent.plugin_manager: successfully launched and dispensed plugin: plugin_name=prometheus
2020-06-10T18:13:18.435-0400 [INFO]  agent.plugin_manager: successfully launched and dispensed plugin: plugin_name=target-value
2020-06-10T18:13:18.528-0400 [INFO]  agent.plugin_manager: successfully launched and dispensed plugin: plugin_name=nomad-target
2020-06-10T18:13:18.529-0400 [INFO]  agent.health_server: server now listening for connections: address=127.0.0.1:9999
2020-06-10T18:13:18.529-0400 [DEBUG] agent.nomad_policy_source: starting policy blocking query watcher
2020-06-10T18:13:18.531-0400 [TRACE] agent.policy_manager: detected 1 policies
2020-06-10T18:13:18.531-0400 [TRACE] agent.policy_manager: creating new handler: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881
2020-06-10T18:13:18.531-0400 [TRACE] policy_handler: starting policy handler: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881
2020-06-10T18:13:18.531-0400 [TRACE] agent.nomad_policy_source: starting policy blocking query watcher: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881
2020-06-10T18:13:18.532-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.533-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.533-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.534-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.535-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.535-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.536-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.536-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.537-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.538-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.538-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.539-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.539-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.540-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.541-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.541-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.542-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.542-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.543-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.544-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.544-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.545-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.546-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.546-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.547-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.547-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.548-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.549-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.549-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
2020-06-10T18:13:18.550-0400 [ERROR] policy_handler: policy validation failed: policy_id=c6346b2e-a381-9391-94b8-7d01ed288881 errors=["scaling.policy.query is missing", "scaling.policy.strategy must be []interface{}, found <nil>"]
...
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -8,6 +8,7 @@ FEATURES:
 
 BUG FIXES:
  * cli: fix incorrect flag help detail for `nomad-ca-path` [[GH-168](https://github.com/hashicorp/nomad-autoscaler/pull/168)]
+ * policy/nomad: fix fast loop when Nomad policy syntax is incorrect [[GH-179](https://github.com/hashicorp/nomad-autoscaler/pull/179)]
 
 ## 0.0.2 (May 21, 2020)
 

--- a/policy/nomad/source.go
+++ b/policy/nomad/source.go
@@ -167,6 +167,11 @@ func (s *Source) MonitorPolicy(ctx context.Context, ID policy.PolicyID, resultCh
 				continue
 			}
 
+			// GH-165: update the wait index. After this point there is a
+			// possibility of continuing the loop and without setting the index
+			// we will just fast loop indefinitely.
+			q.WaitIndex = meta.LastIndex
+
 			if err := validateScalingPolicy(p); err != nil {
 				errMsg := "policy validation failed"
 				if _, ok := err.(*multierror.Error); ok {
@@ -184,7 +189,6 @@ func (s *Source) MonitorPolicy(ctx context.Context, ID policy.PolicyID, resultCh
 			s.canonicalizePolicy(&autoPolicy)
 
 			resultCh <- autoPolicy
-			q.WaitIndex = meta.LastIndex
 		}
 	}
 }

